### ISSUE TITLE
Support for multiple seasonal periods?

### ISSUE THREAD

**Issue by ngupta23** [pycaret/pycaret#1901]
### Discussed in https://github.com/pycaret/pycaret/discussions/1895

<div type='discussions-op-text'>

<sup>Originally posted by **drmario-gh** November 29, 2021</sup>
The `seasonal_period` argument of `setup` only allows to specify a single value. However, several estimators such as TBATS or prophet support multiple seasonalities directly, and I guess the rest could support them via multiple deseasonalizing. I couldn't find anything about this on the roadmap so I was wondering: is there a way to work with multiple seasonal periods in pycaret that I missed?</div>

**Comment by ngupta23** [comment]
@drmario-gh I have created an enhancement request from this discussion but placed it in backlog for now. Would you like to work on this item? If so, we can move it to planned stage.

**Comment by drmario-gh** [comment]
Sure @ngupta23. Not sure how much time I will be able to devote to it on the upcoming days but I will have a go at it.

**Comment by ngupta23** [comment]
Thanks. Let me know if you need any help understanding the code.

**Comment by ngupta23** [comment]
One thing we will need to consider is if multiple seasonalities are passed by user, then which one gets picked for the models that only support one seasonal frequency. Maybe we can do this...

```python
# Picks 7 for all models (currently implemented)
exp.setup(..., seasonal_period = 7) 

# Picks 7 for all models (need to add support for lists)
exp.setup(..., seasonal_period = [7]) 

# Picks 7, 14 for models that support multiple seasonalities. (need to add support for this)
# Picks the first entry (7) for models that only support a single seasonal period
exp.setup(..., seasonal_period = [7, 14]) 
```

Also, we need to understand how we can programmatically determine which model supports a single frequency and which one supports multiple frequencies.

Just some thoughts... feel free to add to it.

**Comment by ngupta23** [comment]
Also, when adding this, please submit a PR to the `time_series` branch for now. Thanks!

**Comment by drmario-gh** [comment]
@ngupta23 I finally had a bit of time to take a look to this. This is what I'm roughly planning to do:
* Will add a `multiple_seasonalities` class attribute to those TimeSeriesContainers that support multiple seasonalities. I will initially start with Prophet and TBATS.
* If multiple_seasonalities are specified when setup experiment, in https://github.com/pycaret/pycaret/blob/14e1a19e811b2a8abdb80f56b9369208d4615a21/pycaret/internal/pycaret_experiment/time_series_experiment.py#L782
I will only select those that allow multiple_seasonalities. I will left as next steps either also picking up the models with single seasonality or even more complex stuff (running an STL decomposition with multiple seasonalities and applying the rest of the models to the seasonally adjusted series).

Does it sound good? Also, what is the average run time of the test suite? It was taking ages on a decent PC. Is it because I need to make use of the GPU there?

**Comment by ngupta23** [comment]
Hi @drmario-gh Thanks for looking into this.

I would recommend a few things:

> Will add a `multiple_seasonalities` class attribute to those TimeSeriesContainers that support multiple seasonalities. I will initially start with Prophet and TBATS.

> If multiple_seasonalities are specified when setup experiment, I will only select those that allow multiple_seasonalities. I will left as next steps either also picking up the models with single seasonality or even more complex stuff (running an STL decomposition with multiple seasonalities and applying the rest of the models to the seasonally adjusted series).

We should introduce a new class attribute to the Experiment Class called `all_sp_values`. This will be a List. In addition, we should rename `sp_to_use` to `primary_sp_to_use`. When there is a single `seasonal_period` value provided or derived from index (e.g. 12), this will have the following

```python
self.all_sp_values = [12]
self.primary_sp_to_use = 12
```

If multiple `seasonal_period` values are provided by the user (say `seasonal_perod = [12, 24]`), we will have 

```python
self.all_sp_values = [12, 24]
self.primary_sp_to_use = 12
```

I still propose that we use the first value in the list as the `primary_sp_to_use` for now. Later, we can implement more intelligent methods to derive `primary_sp_to_use` value automatically from the list (irrespective of order in which they are specified).

The advantage of the above method is that you do not have to disable models if multiple seasonal periods are provided. The ones that only support a single seasonal period will use the `primary_sp_to_use` and those that can support multiple seasonalities can use `all_sp_values`.


> Will add a multiple_seasonalities class attribute to those TimeSeriesContainers that support multiple seasonalities. 

Also, since you already know which model support multiple seasonalities, you can just pick the correct class attribute in the container without adding the `multiple_seasonalities` class attribute in the container. If we really need to add this, this should be added in the `sktime` model itself. You can see, they have several tags associated with properties of the model. 

https://github.com/alan-turing-institute/sktime/blob/b61f44aad75c6ea16e272a5bdd3111b40fc1621f/sktime/forecasting/base/_base.py#L67


> I will initially start with Prophet and TBATS.

I would recommend to prove this out with TBATS first. Once that works, we can extend to other models that support multiple seasonality.

If this sounds good, let me know. I was planning to refactor the setup function shortly so this can be started once that is done.

**Comment by ngupta23** [comment]
> Also, what is the average run time of the test suite? It was taking ages on a decent PC. Is it because I need to make use of the GPU there?

No it should only take about 40-45 minutes to run all the tests. It may be stuck on the plotting tests somewhere. If you run locally, it expects you to close the pop up plot before the test can proceed. If it is a hassle, you can try to run just the time series tests. There are multiple files so you will have to run then one at a time manually unless you know of a way to run multiple files at a time (all of them follow this nomenclature - `test_time_seriees_*.py`).

```
pytest pycaret/tests/test_time_series<full tests name here>.py -v
```

**Comment by drmario-gh** [comment]
Hi @ngupta23,
Sounds cleaner that way, and ok to start with TBATS only. Regarding the testing suite, I was using the [-k flag](https://docs.pytest.org/en/6.2.x/example/markers.html#using-k-expr-to-select-tests-based-on-their-name) to match "time_series" only tests and that was taking more than an hour, but I'll keep an eye to check whether it was simply stuck (didn't see any plot in any case).

**Comment by ngupta23** [comment]
The time series tests take the longest amongst all tests (roughly 1 hour on my machine). Within time series tests. the ones for tuning take the longest (especially `auto_arima`, `bats` and `tbats`). You can try to run one by one to see which one is tripping and we can take it from there.

I also have an open PR that refactors the setup method. It might be better to start once this is merged since you will also be changing the setup method. https://github.com/pycaret/pycaret/pull/2036

**Comment by ngupta23** [comment]
> Regarding the testing suite, I was using the [-k flag](https://docs.pytest.org/en/6.2.x/example/markers.html#using-k-expr-to-select-tests-based-on-their-name) to match "time_series"

Can you provide the entire command? It is not working for me. The tests themselves don't have the word `time_series` in them, just the file name.

**Comment by drmario-gh** [comment]
> The time series tests take the longest amongst all tests (roughly 1 hour on my machine). Within time series tests. the ones for tuning take the longest (especially `auto_arima`, `bats` and `tbats`). You can try to run one by one to see which one is tripping and we can take it from there.

Sure, I didn't tinker with them much, just launched them a couple of times and saw a long run. I'll have a deeper look to this.
 
> I also have an open PR that refactors the setup method. It might be better to start once this is merged since you will also be changing the setup method. #2036
Sure. I'll keep an eye on it.

>Can you provide the entire command? It is not working for me. The tests themselves don't have the word time_series in them, just the file name.

Sure: `pytest -k time_series`. It does not only try to match test names, but also a bunch of other stuff (including filenames). From [pytest docs:](https://docs.pytest.org/en/6.2.x/example/markers.html#using-k-expr-to-select-tests-based-on-their-name) 
>In addition to the test’s name, -k also matches the names of the test’s parents (usually, the name of the file and class it’s in), attributes set on the test function, markers applied to it or its parents and any extra keywords explicitly added to it or its parents.

**Comment by drmario-gh** [comment]
>We should introduce a new class attribute to the Experiment Class called all_sp_values. This will be a List. In addition, we should rename sp_to_use to primary_sp_to_use. When there is a single seasonal_period value provided or derived from index (e.g. 12), this will have the following

Maybe you already had this in mind @ngupta23, but the time series models pick up their seasonality from `seasonal_period` instead of `sp_to_use`, so specifying a list in `seasonal_period` will break models that only support a single seasonality. 
https://github.com/pycaret/pycaret/blob/e994cc04fced16b8d5056fe241076279e11ce473/pycaret/containers/models/time_series.py#L750

Unless I'm missing something, we should probably make the models pick their seasonality from `sp_to_use`. However, that would change existing functionality: `sp_to_use` is 1 if the seasonality specified by the user fails the autocorrelation test. Would we want this?

https://github.com/pycaret/pycaret/blob/e994cc04fced16b8d5056fe241076279e11ce473/pycaret/internal/pycaret_experiment/time_series_experiment.py#L1087

**Comment by ngupta23** [comment]
Good catch. This is a TODO item already. See here: https://github.com/pycaret/pycaret/issues/1894

We want to use `sp_to_use` in the models. And it is ok to use 1 if the seasonal test fails. For example, if we have daily data which is noise, we will first test with seasonal_period = 7. It will fail the test and then we will revert to 1. If multiple seasonal periods are provided by the user, we can try them one by one and stop as soon as we detect seasonality at any value. e.g. if the user specifies `seasonal_period = [7, 14, 28]`, then we first test seasonality at 7. If it fails, we test at 14. If it passes, we set `sp_to_use = 14`. In another case, if all 3 values fail the seasonality test, then we set it `sp_to_use` to 1. What do you think? 

Now that we think about it, it might be better to resolve #1894 before you implement the multiple seasonality. Would you want to pick it up? Should be a simple enough change I hope.

**Comment by ngupta23** [comment]
@drmario-gh Another thing worth noting is that there is an open item about how to handle seasonal_period - whether to use it as is or do a seasonality test first. https://github.com/pycaret/pycaret/issues/1777

Currently, we do the seasonality test and that is ok for the current implementation. In the future, we may provide an option to not enforce the seasonality test and use the values as is.

**Comment by drmario-gh** [comment]
@ngupta23, regarding what to do with multiple seasonalities and the autocorrelation test, yeah, I agree with you, that's what I had in mind. 

>Now that we think about it, it might be better to resolve #1894 before you implement the multiple seasonality. Would you want to pick it up? Should be a simple enough change I hope.

Sure, no problem!

>Currently, we do the seasonality test and that is ok for the current implementation. In the future, we may provide an option to not enforce the seasonality test and use the values as is.
Not much to add there, I agree with what you wrote in #1777. An `enforce_sp` parameter will only have a minor impact in these changes, so we should be good anyway.

**Comment by ngupta23** [comment]
@drmario-gh Do you mind commenting on #1894 so that I can assign it to you? Thanks!

**Comment by drmario-gh** [comment]
Hi @ngupta23,

- I've finally decided to leave the renaming of `sp_to_use` for a different PR to make reviews easier.
- Shouldn't we allow non integer seasonal values too (where it makes sense)? E.g., TBATS and Prophet support them.

**Comment by ngupta23** [comment]
> I've finally decided to leave the renaming of sp_to_use for a different PR to make reviews easier.

That is a good idea. We are also undergoing a major refactor of the code which will get merged into the `time_series` branch shortly. We should wait for that to be done before we do this changes to make it easier.

> Shouldn't we allow non integer seasonal values too (where it makes sense)? E.g., TBATS and Prophet support them.

Good idea. If the model can support it, we can provide it, but we have to coordinate this appropriately. For example, if the user just specifies seasonal_period = 52.14, how do we handle this in models that do not take floating values. Maybe round it for them? It probably warrants a separate Issue where we can plan and discuss this.

**Comment by ngupta23** [comment]
Hi @drmario-gh Thank you for closing out this development. This will be a good addition.

I have created 2 new issues for the ones you pointed out. Let me know if you would like to pick up any of these. 

Thanks again!

- #2165 
- #2166
